File: presentation/src/main/java/com/github/anrimian/musicplayer/ui/editor/composition/CompositionEditorView.java
package com.github.anrimian.musicplayer.ui.editor.composition;

import com.github.anrimian.musicplayer.domain.models.composition.FullComposition;
import com.github.anrimian.musicplayer.domain.models.genres.ShortGenre;
import com.github.anrimian.musicplayer.ui.common.error.ErrorCommand;
import com.github.anrimian.musicplayer.ui.utils.moxy.SingleStateByTagStrategy;

import java.util.List;

import moxy.MvpView;
import moxy.viewstate.strategy.AddToEndSingleStrategy;
import moxy.viewstate.strategy.OneExecutionStateStrategy;
import moxy.viewstate.strategy.SkipStrategy;
import moxy.viewstate.strategy.StateStrategyType;

public interface CompositionEditorView extends MvpView {

    String DISPLAY_COMPOSITION_STATE = "display_composition_state";

    @StateStrategyType(OneExecutionStateStrategy.class)
    void closeScreen();

    @StateStrategyType(value = SingleStateByTagStrategy.class, tag = DISPLAY_COMPOSITION_STATE)
    void showCompositionLoadingError(ErrorCommand errorCommand);

    @StateStrategyType(value = SingleStateByTagStrategy.class, tag = DISPLAY_COMPOSITION_STATE)
    void showComposition(FullComposition composition);

    @StateStrategyType(OneExecutionStateStrategy.class)
    void showErrorMessage(ErrorCommand errorCommand);

    @StateStrategyType(SkipStrategy.class)
    void showEnterAuthorDialog(FullComposition composition, String[] hints);

    @StateStrategyType(SkipStrategy.class)
    void showEnterTitleDialog(FullComposition composition);

    @StateStrategyType(SkipStrategy.class)
    void showEnterFileNameDialog(FullComposition composition);

    @StateStrategyType(SkipStrategy.class)
    void copyFileNameText(String filePath);

    @StateStrategyType(SkipStrategy.class)
    void showEnterAlbumDialog(FullComposition composition, String[] hints);

    @StateStrategyType(SkipStrategy.class)
    void showEnterAlbumArtistDialog(FullComposition composition, String[] hints);

    @StateStrategyType(SkipStrategy.class)
    void showAddGenreDialog(String[] genres);

    @StateStrategyType(SkipStrategy.class)
    void showEditGenreDialog(ShortGenre shortGenre, String[] genres);

    @StateStrategyType(AddToEndSingleStrategy.class)
    void showGenres(List<ShortGenre> shortGenres);

    @StateStrategyType(OneExecutionStateStrategy.class)
    void showRemovedGenreMessage(ShortGenre genre);

    @StateStrategyType(SkipStrategy.class)
    void showCoverActionsDialog();
}

File: presentation/src/main/java/com/github/anrimian/musicplayer/ui/editor/composition/CompositionEditorView.kt
package com.github.anrimian.musicplayer.ui.editor.composition;

import com.github.anrimian.musicplayer.domain.models.composition.FullComposition;
import com.github.anrimian.musicplayer.domain.models.genres.ShortGenre;
import com.github.anrimian.musicplayer.ui.common.error.ErrorCommand;
import com.github.anrimian.musicplayer.ui.utils.moxy.SingleStateByTagStrategy;

import java.util.List;

import moxy.MvpView;
import moxy.viewstate.strategy.StateStrategyType;
import moxy.viewstate.strategy.alias.AddToEndSingle;
import moxy.viewstate.strategy.alias.OneExecution;
import moxy.viewstate.strategy.alias.Skip;

public interface CompositionEditorView extends MvpView {

    String DISPLAY_COMPOSITION_STATE = "display_composition_state";
    String CHANGE_FILE_STATE = "change_file_state";

    @OneExecution
    void closeScreen();

    @StateStrategyType(value = SingleStateByTagStrategy.class, tag = DISPLAY_COMPOSITION_STATE)
    void showCompositionLoadingError(ErrorCommand errorCommand);

    @StateStrategyType(value = SingleStateByTagStrategy.class, tag = DISPLAY_COMPOSITION_STATE)
    void showComposition(FullComposition composition);

    @StateStrategyType(value = SingleStateByTagStrategy.class, tag = CHANGE_FILE_STATE)
    void showChangeFileProgress();

    @StateStrategyType(value = SingleStateByTagStrategy.class, tag = CHANGE_FILE_STATE)
    void hideChangeFileProgress();

    @OneExecution
    void showErrorMessage(ErrorCommand errorCommand);

    @OneExecution
    void showCheckTagsErrorMessage(ErrorCommand errorCommand);

    @Skip
    void showEnterAuthorDialog(FullComposition composition, String[] hints);

    @Skip
    void showEnterTitleDialog(FullComposition composition);

    @Skip
    void showEnterFileNameDialog(FullComposition composition);

    @Skip
    void copyFileNameText(String filePath);

    @Skip
    void showEnterAlbumDialog(FullComposition composition, String[] hints);

    @Skip
    void showEnterAlbumArtistDialog(FullComposition composition, String[] hints);

    @Skip
    void showEnterLyricsDialog(FullComposition composition);

    @Skip
    void showAddGenreDialog(String[] genres);

    @Skip
    void showEditGenreDialog(ShortGenre shortGenre, String[] genres);

    @AddToEndSingle
    void showGenres(List<ShortGenre> shortGenres);

    @OneExecution
    void showRemovedGenreMessage(ShortGenre genre);

    @Skip
    void showCoverActionsDialog();

    @Skip
    void showSelectImageFromGalleryScreen();
}
